dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp

```diff
@@ -13,19 +13,496 @@
+{
+    if unlikely (resp.has_error())
+    {
+        handleBackgroundError(resp.error().message());
+        return {};
+    }
+
+    std::vector<std::string> handled_resource_group_names;
+    handled_resource_group_names.reserve(resp.responses_size());
+    // Network to GAC is ok, update timepoint.
+    last_fetch_tokens_from_gac_timepoint = std::chrono::steady_clock::now();
+
+    if (resp.responses().empty())
+    {
+        handleBackgroundError("got empty TokenBuckets resp from GAC");
+        return {};
+    }
+
+    const auto now = std::chrono::steady_clock::now();
+    for (const resource_manager::TokenBucketResponse & one_resp : resp.responses())
+    {
+        // For each resource group.
+        if unlikely (!one_resp.granted_resource_tokens().empty())
+        {
+            handleBackgroundError("GAC return RAW granted tokens, but LAC expect RU tokens");
+            continue;
+        }
+
+        handled_resource_group_names.emplace_back(one_resp.resource_group_name());
+
+        if unlikely (one_resp.granted_r_u_tokens().size() != 1)
+        {
+            handleBackgroundError(fmt::format(
+                "expect resp.granted_r_u_tokens().size() is 1, but got {}",
+                one_resp.granted_r_u_tokens().size()));
+            continue;
+        }
+        auto resource_group = findResourceGroup(one_resp.resource_group_name());
+        if (resource_group == nullptr)
+            continue;
 
-LocalAdmissionController::LocalAdmissionController()
+        const resource_manager::GrantedRUTokenBucket & granted_token_bucket = one_resp.granted_r_u_tokens()[0];
+        if unlikely (granted_token_bucket.type() != resource_manager::RequestUnitType::RU)
+        {
+            handleBackgroundError("unexpected request type");
+            continue;
+        }
+
+        int64_t trickle_ms = granted_token_bucket.trickle_time_ms();
+        RUNTIME_CHECK(trickle_ms >= 0);
+
+        double added_tokens = granted_token_bucket.granted_tokens().tokens();
+        RUNTIME_CHECK(added_tokens >= 0);
+
+        int64_t capacity = granted_token_bucket.granted_tokens().settings().burst_limit();
+
+        // fill_rate should never be setted.
+        RUNTIME_CHECK(granted_token_bucket.granted_tokens().settings().fill_rate() == 0);
+
+        if (trickle_ms == 0)
+        {
+            // GAC has enough tokens for LAC.
+            resource_group->updateNormalMode(added_tokens, capacity);
+        }
+        else
+        {
+            // GAC doesn't have enough tokens for LAC, start to trickle.
+            resource_group->updateTrickleMode(added_tokens, capacity, trickle_ms);
+        }
+        resource_group->updateFetchTokenTimepoint(now);
+    }
+    return handled_resource_group_names;
+}
+
+void LocalAdmissionController::watchGAC()
 {
-    if (!global_instance)
-        global_instance = std::make_unique<MockLocalAdmissionController>();
+    while (true)
+    {
+        doWatch();
```

Comment:
Is a catch exception needed here, could the thread exit due to a network error?

Reply:
I think grpc use `status` as error handling. But it's ok to catch and log here.

Reply:
done